mlflow/projects/__init__.py

```diff
@@ -251,46 +264,127 @@ def _maybe_create_conda_env(conda_env_path):
                           conda_env_path], stream_output=True)
 
 
-def _launch_local_run(active_run, command, work_dir, env_map, stream_output):
+def _launch_local_run(
+        uri, run_id, entry_point, parameters, work_dir, env_map, use_conda, storage_dir,
+        stream_output):
     """
     Runs an entry point by launching its command in a subprocess, updating the tracking server with
     the run's exit status.
 
-    :param active_run: `ActiveRun` to which to post status updates for the launched run
     :param command: Entry point command to execute
     :param work_dir: Working directory to use when executing `command`
     :param env_map: Dict of environment variable key-value pairs to set in the process for `command`
     :return `SubmittedRun` corresponding to the launched run.
     """
-    from mlflow.projects.pollable_run import LocalPollableRun
-    pollable_run = LocalPollableRun(
-        command=command, work_dir=work_dir, env_map=env_map, stream_output=stream_output)
-    return SubmittedRun(active_run, pollable_run)
+    from mlflow.projects.submitted_run import LocalSubmittedRun
+    command = _load_project(work_dir, uri)
+    mlflow_run_arr = ["mlflow", "run", work_dir, "-e", entry_point, "-m", MLFLOW_ENTRY_POINT_MODE]
+    if storage_dir is not None:
+        mlflow_run_arr.extend(["--storage-dir", storage_dir])
+    if not use_conda:
+        mlflow_run_arr.append("--no-conda")
+    if run_id:
+        mlflow_run_arr.extend(["--run-id", run_id])
+    if parameters:
+        for key, value in parameters.items():
+            mlflow_run_arr.extend(["-P", "%s=%s" % (key, value)])
+    final_env = os.environ.copy()
+    final_env.update(env_map)
+    if stream_output:
+        popen = subprocess.Popen(
+            mlflow_run_arr, cwd=work_dir, env=final_env, universal_newlines=True)
+    else:
+        popen = subprocess.Popen(
+            mlflow_run_arr, cwd=work_dir, env=final_env, universal_newlines=True,
+            stderr=open(os.devnull, "w"), stdout=open(os.devnull, "w"))
+    return LocalSubmittedRun(run_id, popen, command)
 
 
-def _run_project(project, entry_point, work_dir, parameters, use_conda, storage_dir,
-                 experiment_id, block):
-    """Locally run a project that has been checked out in `work_dir`."""
+def _maybe_set_run_terminated(active_run, status):
+    """
+    If the passed-in active run is defined and still running (i.e. hasn't already been terminated
+    within user code), mark it as terminated with the passed-in status.
+    """
+    if active_run and not RunStatus.is_terminated(active_run.get_run().info.status):
+        active_run.set_terminated(status)
+
+
+def _run_local_noenv(
```

Comment:
So all this code in `_run_local_noenv` and `_run_entry_point_command` is specific to the case where we are doing `projects.run`, where the mode is local? Is that true?